storage/aptosdb/src/pruner/worker.rs

```diff
@@ -0,0 +1,140 @@
+                match result {
+                    Ok(_) => {}
+                    Err(_) => {
+                        // On error, stop retrying vigorously by making next recv() blocking.
+                        self.blocking_recv = true;
+                    }
+                }
+            }
+            for db_pruner in &self.db_pruners {
+                // if any of the pruner has pending pruning, then we don't block on receive
+                if db_pruner.lock().is_pruning_pending() {
+                    self.blocking_recv = false
+                }
+            }
+            self.record_progress();
+        }
+    }
+
+    fn record_progress(&mut self) {
+        let mut updated_least_readable_versions: Vec<Version> = Vec::new();
+        for x in &self.db_pruners {
+            updated_least_readable_versions.push(x.lock().least_readable_version())
+        }
+        *self.least_readable_versions.lock() = updated_least_readable_versions;
+    }
+
+    /// Tries to receive all pending commands, blocking waits for the next command if no work needs
+    /// to be done, otherwise quits with `true` to allow the outer loop to do some work before
+    /// getting back here.
+    ///
+    /// Returns `false` if `Command::Quit` is received, to break the outer loop and let
+    /// `work_loop()` return.
+    fn receive_commands(&mut self) -> bool {
+        loop {
+            let command = if self.blocking_recv {
+                // Worker has nothing to do, blocking wait for the next command.
+                self.command_receiver
+                    .recv()
+                    .expect("Sender should not destruct prematurely.")
+            } else {
+                // Worker has pending work to do, non-blocking recv.
+                match self.command_receiver.try_recv() {
+                    Ok(command) => command,
+                    // Channel has drained, yield control to the outer loop.
+                    Err(_) => return true,
+                }
+            };
+
+            match command {
+                // On `Command::Quit` inform the outer loop to quit by returning `false`.
+                Command::Quit => return false,
+                Command::Prune { target_db_versions } => {
+                    for (new_target_version, pruner) in
+                        zip_eq(&target_db_versions, &self.db_pruners)
+                    {
+                        if *new_target_version > pruner.lock().target_version() {
+                            // Switch to non-blocking to allow some work to be done after the
+                            // channel has drained.
+                            self.blocking_recv = false;
+                        }
+                        pruner.lock().set_target_version(*new_target_version);
+                    }
+                }
+            }
+        }
+    }
+}
+
+pub enum Command {
+    Quit,
+    Prune { target_db_versions: Vec<Version> },
```

Comment:
Can use tuple instead of `Vec`

Reply:
Yeah that makes sense. (Again, I thought there are only two targets :P )

It's another place to modify when you introduce more pruners.. which will be fine, but did you think around actually "group" those ledger history pruners? Or a LedgerHistory pruner and sub pruners governed by it?

Reply:
Let me find a cleaner way to do this. We can have one place where we could add all the pruners - Is it okay if we do this as a follow up though?

Reply:
For sure. It can also be a further improvement to make after you land your addition of more pruners PR. I don't want to mess up your stack.